middleware/get_head.go

```diff
@@ -0,0 +1,37 @@
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/go-chi/chi"
+)
+
+func GetHead(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == "HEAD" {
+			rctx := chi.RouteContext(r.Context())
+			routePath := rctx.RoutePath
+			if routePath == "" {
+				if r.URL.RawPath != "" {
+					routePath = r.URL.RawPath
+				} else {
+					routePath = r.URL.Path
+				}
+			}
+
+			// Temporary routing context to look-ahead before routing the request
+			tctx := chi.NewRouteContext()
+			tctx.RouteMethod = "HEAD"
+			tctx.RoutePath = routePath
+
+			h := rctx.Routes.Find(tctx, tctx.RouteMethod, tctx.RoutePath)
```

Comment:
Just thinking out loud -- if this method finds a corresponding `handler`, how can we invoke it?

And if we can invoke it, would be its middleware chain invoked too or not?

Reply:
Currently the handler found is the endpoint handler, not the computed mux handler which contains the middlewares. The online middlewares part of the endpoint handler are the inline handlers which come from `With()` or `Group()` middlewares.

Reply:
So basically we verify that the combination of RoutePath and RoutheMethod does have an endpoint, but we cannot call it because we are not sure that this route will go through all setup middlewares.

I have to ask - can we use the returned http.Handler in any meaningful way? If not - maybe we should return bool and rename it to IsExist or something like that.

Reply:
yea, I had a similar thought. Initially the handler was useful, but after writing the first implementation and realizing the need to traverse subrouters and consider explicit HEAD handlers in the tree for some subrouters and not others, the handler return value isn't very useful. It's possible to reconstruct the discrete path, or a computed path, but Id rather find a real need for that with this method, because that is what ServeHTTP() is for. A `bool` is better for checking if the path can find a matching route in the tree.

Reply:
I agree. `rctx.Routes.Exists(tctx, method, type) bool` sounds good to me, since we can't use the handler anyway.

Reply:
```go
tctx := chi.NewRouteContext()
tctx.RouteMethod = "HEAD"
tctx.RoutePath = routePath

h := rctx.Routes.Find(tctx, tctx.RouteMethod, tctx.RoutePath)
```

Seems like the `tctx` has both Method and Route defined. But these two values are passed to the `.Find()` method as arguments again.

That leads to a question:
1. Can we create `tctx` within the `.Find()` method? (ie. `h := rctx.Routes.Find("HEAD", routePath)`)
2. Or can we just pass the context and not the two duplicated `tctx.RouteMethod`, `tctx.RoutePath` arguments? (ie. `h := rctx.Routes.Find(tctx)`)

Reply:
Since we modify the rctx after we had used both RouteMethod and RoutePath - it may be a good idea just to use rctx. 

Hoverwer - if we want a clean API, I think that we can move the Find method body into unexported find method and just call it from the public one. In public one, we will create hte needed rctx.

Reply:
But, from the API viewpoint - maybe we could add a new private interface type and use it to cleanup API - this way we can achieve the similar performance without mental overhead for the end user.

Reply:
Something like
```go
// Find searches the routing tree for a handler that matches the method/path.
// It's similar to routing a http request, but without executing the handler
// thereafter.
func (mx *Mux) Find(method, path string) http.Handler {
	tctx := NewRouteContext()
	tctx.RouteMethod = method
	tctx.RoutePath = path
	return mx.find(tctx, method, path)
}

func (mx *Mux) find(rctx *Context, method, path string) http.Handler {
	m, ok := methodMap[method]
	if !ok {
		return nil
	}

	node, _, h := mx.tree.FindRoute(rctx, m, path)

	if node != nil && node.subroutes != nil {
		rctx.RoutePath = mx.nextRoutePath(rctx)
		if subroute, ok := node.subroutes.(extendedRoutes); ok {
			// Our type is implementing allocation free find. Use it.
			return subroute.find(rctx, rctx.RouteMethod, rctx.RoutePath)
		}
		// Fallback to the public method
		return node.subroutes.Find(rctx.RouteMethod, rctx.RoutePath)
	}

	return h
}
```

Where 
```go
// Routes interface adds two methods for router traversal, which is also
// used by the `docgen` subpackage to generation documentation for Routers.
type Routes interface {
	// Routes returns the routing tree in an easily traversable structure.
	Routes() []Route

	// Middlewares returns the list of middlewares in use by the router.
	Middlewares() Middlewares

	// Find searches the routing tree for a handler that matches
	// the method/path - similar to routing a http request, but without
	// executing the handler thereafter.
	Find(method, path string) http.Handler
}

type extendedRoutes interface {
	Routes
	find(rctx *Context, method, path string) http.Handler
}
```

Reply:
There are advantages to passing the chi.Context yourself, since you can look at the URI parameters that were parsed from the route. But also, you could pool them if you really wanted to.

Reply:
> But also, you could pool them if you really wanted to.

Sorry - can you elaborate?

Reply:
In the case you're creating a lot of NewRouteContext() objects and finding it to cause memory pressure for whatever reason (which I do not expect to be an issue), you could do something similar as here: https://github.com/go-chi/chi/blob/master/mux.go#L51-L53 - that is, make a bunch of temporary routing contexts and grab them from the pool. But, its not really necessary for just matching